src/MiddlewareListener.php

```diff
@@ -64,7 +80,15 @@ public function onDispatch(MvcEvent $event)
             foreach ($routeMatch->getParams() as $key => $value) {
                 $psr7Request = $psr7Request->withAttribute($key, $value);
             }
-            $return = $middleware($psr7Request, Psr7Response::fromZend($response));
+            $return = $pipe(
+                $psr7Request,
+                $psr7ResponsePrototype,
+                function (PsrServerRequestInterface $request, PsrResponseInterface $response) {
+                    throw new ReachedFinalHandlerException(
```

Comment:
One potential refactor I see here is to instead define `ReachedFinalHandlerException` as follows:

```php
class ReachedFinalHandlerException extends RuntimeException
{
    public static function createFinalHandler()
    {
        return new self('Reached the final handler for middleware pipe - check the pipe configuration');
    }

    public function __invoke()
    {
        throw $this;
    }
}
```

This would then allow you to do the following above:

```php
$return = $pipe($psr7Request, $psr7Response, ReachedFinalHandlerException::createFinalHandler());
```

Not sure if that's abusing inheritance and SRP, though. :grin: